--- a/src/connections.c
+++ b/src/connections.c
@@ -1248,6 +1248,7 @@ static handler_t connection_handle_fdeve
 	if (con->state == CON_STATE_READ ||
 	    con->state == CON_STATE_READ_POST) {
 		connection_handle_read_state(srv, con);
+		if (con->is_readable) joblist_append(srv, con);
 	}
 
 	if (con->state == CON_STATE_WRITE &&
@@ -1653,6 +1654,7 @@ int connection_state_machine(server *srv
 			}
 
 			connection_handle_read_state(srv, con);
+			if (con->is_readable) joblist_append(srv, con);
 			break;
 		case CON_STATE_WRITE:
 			if (srv->srvconf.log_state_handling) {
